### PR TITLE
[Issue #57] Implement Team City retest failed functionality

### PR DESCRIPTION
**Pull request by khermano**
Issue: https://github.com/jboss/tyr/issues/57

### CODE REVIEW

File: tyr-runner/src/main/java/org/jboss/tyr/ci/ContinuousIntegration.java [hunk 1]
```diff
@@ -19,9 +19,9 @@
 
 public interface ContinuousIntegration {
 
-    void triggerBuild(JsonObject prPayload);
+    void triggerRetesting(JsonObject prPayload);
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/ContinuousIntegration.java]: what is wrong with the original?

File: tyr-runner/src/main/java/org/jboss/tyr/ci/ContinuousIntegration.java [hunk 2]
```diff
@@ -19,9 +19,9 @@
...
 
-    void triggerFailedBuild(JsonObject prPayload);
+    void triggerRetestingFailed(JsonObject prPayload);
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/ContinuousIntegration.java]:
```suggestion
    void triggerFailedTests(JsonObject prPayload);
```

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 3]
```diff
@@ -101,29 +124,100 @@ public void triggerBuild(JsonObject prPayload) {
...
-    private String encryptCredentials(String username, String password) {
-        String authStr = username + ":" + password;
-        return DatatypeConverter.printBase64Binary(authStr.getBytes(StandardCharsets.UTF_8));
+    @Override
+    public void triggerRetestingFailed(JsonObject prPayload) {
+        setTriggerBuildProperties(prPayload);
+        if (!branchMappings.containsKey(branch)) {
+            return;
+        }
+        setSnapshotDependencies(buildId, true);
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: Can you refactor this somehow please? This method and triggerRetesting differ only in the boolean value passed to this invocation.

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 4]
```diff
@@ -101,29 +124,100 @@ public void triggerBuild(JsonObject prPayload) {
...
 
-        for (String mapping : mappings.split(",")) {
-            String[] parts = mapping.split("=>");
-            branchMap.put(parts[0].trim(), parts[1].trim());
+        try {
+            response = target.request()
+                .header(HttpHeaders.ACCEPT_ENCODING, "UTF-8")
+                .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON)
+                .header(HttpHeaders.AUTHORIZATION, "Basic " + encryptedCredentials)
+                .header(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON)
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: maybe you can return somewhere this whole request builder object so you don't need to repeat all the headers everytime?

Comment (xstefank) [reply tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: or even more generic `getJson(path)`

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 5]
```diff
@@ -101,29 +124,100 @@ public void triggerBuild(JsonObject prPayload) {
...
         }
+        return snapshotDependencies;
+    }
 
-        return branchMap;
+    private void setTriggerBuildProperties(JsonObject prPayload) {
+        branch = prPayload.getJsonObject(Utils.BASE).getString(Utils.REF);
+        pull = prPayload.getInt(Utils.NUMBER);
+        sha = prPayload.getJsonObject(Utils.HEAD).getString(Utils.SHA);
+        buildId = branchMappings.get(branch);
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: so you are setting these fields on every invocation of triggerRetesting*. This kind of storing values is inefficient. If they are stored only for one method invocation then it is not necessary to store them in fields.

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 6]
```diff
@@ -101,29 +124,100 @@ public void triggerBuild(JsonObject prPayload) {
...
     }
 
-    @Override
-    public void triggerFailedBuild(JsonObject prPayload) {
-        throw new UnsupportedOperationException("Method is not implemented");
+    private WebTarget returnTeamCityTarget(String path) {
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]:
```suggestion
    private WebTarget getTeamCityTarget(String path) {
```

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 7]
```diff
@@ -101,29 +124,100 @@ public void triggerBuild(JsonObject prPayload) {
...
+        Client client = ClientBuilder.newClient();
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: client object can be reused

File: tyr-runner/src/main/java/org/jboss/tyr/model/json/Property.java [hunk 8]
```diff
@@ -15,11 +15,18 @@
  */
 package org.jboss.tyr.model.json;
 
+import javax.json.bind.annotation.JsonbPropertyOrder;
+
+@JsonbPropertyOrder({"name", "value"})
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/model/json/Property.java]: why is this important, please?

File: tyr-runner/src/main/java/org/jboss/tyr/model/json/SnapshotDependencies.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
+package org.jboss.tyr.model.json;
+
+import javax.json.bind.annotation.JsonbProperty;
+import javax.json.bind.annotation.JsonbPropertyOrder;
+import java.math.BigDecimal;
+import java.util.List;
+
+@JsonbPropertyOrder({"count", "snapshot-dependency"})
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/model/json/SnapshotDependencies.java]: why is this important, please?

File: tyr-runner/src/main/java/org/jboss/tyr/model/json/SnapshotDependency.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
+package org.jboss.tyr.model.json;
+
+import javax.json.bind.annotation.JsonbProperty;
+import javax.json.bind.annotation.JsonbPropertyOrder;
+
+@JsonbPropertyOrder({"id", "type", "properties", "source-buildType"})
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/model/json/SnapshotDependency.java]: sorry but again why is the order important?

File: tyr-runner/src/main/java/org/jboss/tyr/ci/ContinuousIntegration.java (hunk 1 above)
Comment (khermano) [reply tyr-runner/src/main/java/org/jboss/tyr/ci/ContinuousIntegration.java]: There still is the original method https://github.com/khermano/tyr/commit/0bbcf1a3ad81bba448f8b38d878b73449c539915#diff-2495a7c9254abaaeb22b96823ce7bb5aR103 .... I just want to separate its functionality because of DRY principle as I use this method in both triggerRetesting() and triggerRetestingFailed() methods.  :)

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java (hunk 3 above)
Comment (khermano) [reply tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: Yes, of course. :)

File: tyr-runner/src/main/java/org/jboss/tyr/ci/ContinuousIntegration.java (hunk 1 above)
Comment (xstefank) [reply tyr-runner/src/main/java/org/jboss/tyr/ci/ContinuousIntegration.java]: that is a private method in the implementation. That's not a reason to rename interface.

Comment (khermano) [reply tyr-runner/src/main/java/org/jboss/tyr/ci/ContinuousIntegration.java]: Sure, I understand your point. I will rename it back and change name of the private method. :)

File: tyr-runner/src/main/java/org/jboss/tyr/model/json/Property.java (hunk 8 above)
Comment (khermano) [reply tyr-runner/src/main/java/org/jboss/tyr/model/json/Property.java]: It is not anymore...there was some problem with mapping but that is irrelevant now...I will remove it. Sorry, my bad. :) This applies to all @JsonbPropertyOrder annotation which I added.

**Comment by khermano** [comment]
Messages are flowing between Tyr and TeamCity:

- Tyr receives PR comment JSON
- trigger* method collects necessary info from upper JSON and config.properties file
- Tyr invokes e.g. `http://localhost:8111/app/rest/buildTypes/id:TestRepo_Aggregator/snapshot-dependencies` to get JSON with snapshot dependencies (contains build configurations from build chain) of aggregation build configuration (buildId = TestRepo_Aggregator)
- Tyr put same JSON with snapshot dependencies on the same URL changing only `take-started-build-with-same-revisions` and `take-successful-builds-only` properties in every snapshot dependency depending on whether we want run all builds or only failed ones (true if we want to test only failed ones)
- Tyr triggers build by posting build JSON to e.g. `http://localhost:8111/app/rest/buildQueue`

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 11]
```diff
@@ -73,57 +110,103 @@ public void triggerBuild(JsonObject prPayload) {
...
-    }
+        if (snapshotDependencies.snapshotDependencyCount.intValue() < 1) {
+            log.info("No snapshot dependency.");
+            return;
+        }
 
-    private String getBaseUrl(String host, int port) {
-        return String.format("http://%s:%d/httpAuth", host, port);
+        for (SnapshotDependency dependency: snapshotDependencies.snapshotDependencies) {
+            for (Property prop :dependency.properties.propertyList) {
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: just minor code style preference, use space around ':' on both sides in for-each loops please

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 12]
```diff
@@ -73,57 +110,103 @@ public void triggerBuild(JsonObject prPayload) {
...
+                if (prop.name.equals("take-started-build-with-same-revisions") || prop.name.equals("take-successful-builds-only")) {
+                    prop.value = String.valueOf(retestOnlyFailed);
+                }
+            }
+        }
+
+        Jsonb jsonb = JsonbBuilder.create();
+        String stringJsonb = jsonb.toJson(snapshotDependencies);
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: it's not a string with jsonb right? it's string with snapshotDependencies :)

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 13]
```diff
@@ -73,57 +110,103 @@ public void triggerBuild(JsonObject prPayload) {
...
+        Entity<String> json = Entity.json(stringJsonb);
+
+        String path = "app/rest/buildTypes/id:" + buildId + "/snapshot-dependencies";
+        WebTarget target = getTeamCityTarget(path);
+        Response response = null;
+
+        try {
+            response = getHeaderBuilder(target).put(json);
+            log.info("Putting JSON with snapshot dependencies in Teamcity. Status: " + response.getStatus());
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: Sending JSON ...

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 14]
```diff
@@ -73,57 +110,103 @@ public void triggerBuild(JsonObject prPayload) {
...
+        WebTarget target = getTeamCityTarget(path);
+        Response response = null;
+        SnapshotDependencies snapshotDependencies;
 
-        for (String mapping : mappings.split(",")) {
-            String[] parts = mapping.split("=>");
-            branchMap.put(parts[0].trim(), parts[1].trim());
+        try {
+            response = getHeaderBuilder(target).get();
+            log.info("Getting JSON with snapshot dependencies from Teamcity. Status: " + response.getStatus());
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: use debug for these log messages please (also putting... above)

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 15]
```diff
@@ -73,57 +110,103 @@ public void triggerBuild(JsonObject prPayload) {
...
+
+            String json = response.readEntity(String.class);
+            Jsonb jsonb = JsonbBuilder.create();
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: you can cache Jsonb object in a field

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 16]
```diff
@@ -73,57 +110,103 @@ public void triggerBuild(JsonObject prPayload) {
...
+        } finally {
+            if (response != null) {
+                response.close();
+            }
         }
+        return snapshotDependencies;
+    }
 
-        return branchMap;
+    private Invocation.Builder getHeaderBuilder(WebTarget target) {
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]:
```suggestion
    private Invocation.Builder getRequestBuilder(WebTarget target) {
```

File: tyr-runner/src/main/java/org/jboss/tyr/model/json/Property.java [hunk 17]
```diff
@@ -20,6 +20,10 @@
     public String name;
     public Object value;
 
+    public Property() {
+
+    }
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/model/json/Property.java]: why is this needed?

File: tyr-runner/src/main/java/org/jboss/tyr/model/json/SnapshotDependencies.java [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+
+    @JsonbProperty("count")
+    public BigDecimal snapshotDependencyCount;
+
+    @JsonbProperty("snapshot-dependency")
+    public List<SnapshotDependency> snapshotDependencies;
+
+    public SnapshotDependencies() {
+
+    }
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/model/json/SnapshotDependencies.java]: also, why is the default constructor not enough? Also in other Json classes. Did you try to run the app without these empty constructors?

File: tyr-runner/src/test/java/org/jboss/tyr/ci/TeamCityCITest.java [hunk 19]
```diff
@@ -15,12 +15,55 @@
...
+    @Test
+    public void testTriggerFailedTestsNullParameters() {
+        exception.expect(NullPointerException.class);
+        new TeamCityCI().triggerFailedTests(null);
+    }
+
+    @Test
+    public void testTriggerBuildPrPayloadMissingPullRequestParameter() {
+        exception.expect(NullPointerException.class);
+        new TeamCityCI().triggerBuild(TestUtils.EMPTY_PAYLOAD);
```
Comment (xstefank) [inline tyr-runner/src/test/java/org/jboss/tyr/ci/TeamCityCITest.java]: shouldn't this raise something like "Invalid payload exception"? Also for other tests after this one.

File: tyr-runner/src/main/java/org/jboss/tyr/model/json/Property.java (hunk 17 above)
Comment (khermano) [reply tyr-runner/src/main/java/org/jboss/tyr/model/json/Property.java]: cause I am reusing this already existing class for jsonb unmarshalling and it requires noargconstructor :) when I remove it I am getting NPE here https://github.com/khermano/tyr/commit/9fe776e81c5361c482c3abf7bda3e18fc90b6b5e#diff-2495a7c9254abaaeb22b96823ce7bb5aR183

File: tyr-runner/src/main/java/org/jboss/tyr/model/json/SnapshotDependencies.java (hunk 18 above)
Comment (khermano) [reply tyr-runner/src/main/java/org/jboss/tyr/model/json/SnapshotDependencies.java]: there you are right :) these I don't need because of default constructor...sorry for that, they are only remains when I was trying to use parametric constructor too..removing them :)

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 20]
```diff
@@ -60,37 +67,69 @@ public void init() {
...
+            pull = prPayload.getInt(Utils.NUMBER);
+            sha = prPayload.getJsonObject(Utils.HEAD).getString(Utils.SHA);
+            buildId = branchMappings.get(branch);
+        } catch (NullPointerException e) {
+            throw new InvalidPayloadException("Invalid payload, can't retrieve all elements. Message was: " + e.getMessage());
+        }
+
+        setSnapshotDependencies(buildId, retestOnlyFailed);
+
+        String path = "/app/rest/buildQueue";
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: this could be a constant for easier access, but not critical

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 21]
```diff
@@ -101,29 +140,81 @@ public void triggerBuild(JsonObject prPayload) {
...
+        if (snapshotDependencies.snapshotDependencyCount.intValue() < 1) {
+            log.info("No snapshot dependency.");
+            return;
+        }
 
-    private String getBaseUrl(String host, int port) {
-        return String.format("http://%s:%d/httpAuth", host, port);
+        for (SnapshotDependency dependency : snapshotDependencies.snapshotDependencies) {
+            for (Property prop : dependency.properties.propertyList) {
+                if (prop.name.equals("take-started-build-with-same-revisions") || prop.name.equals("take-successful-builds-only")) {
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: also better in constants

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 22]
```diff
@@ -101,29 +140,81 @@ public void triggerBuild(JsonObject prPayload) {
...
+                    prop.value = String.valueOf(retestOnlyFailed);
+                }
+            }
+        }
+
+        Jsonb jsonb = JsonbBuilder.create();
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: better to be cached. No need to create a new instance on every method call.

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 23]
```diff
@@ -101,29 +140,81 @@ public void triggerBuild(JsonObject prPayload) {
...
+
+        String path = "app/rest/buildTypes/id:" + buildId + "/snapshot-dependencies";
+        WebTarget target = getTeamCityTarget(path);
+        Response response = null;
+
+        try {
+            response = getRequestBuilder(target).put(json);
+            log.debug("Sending JSON with snapshot dependencies in Teamcity. Status: " + response.getStatus());
+        } catch (Throwable e) {
+            log.error("Cannot retrieve JSON from " + target.getUri(), e);
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: you are sending JSON right? :)

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 24]
```diff
@@ -101,29 +140,81 @@ public void triggerBuild(JsonObject prPayload) {
...
+            if (response != null) {
+                response.close();
+            }
+        }
     }
 
-    private Map<String, String> parseBranchMapping(String mappings) {
-        Map<String, String> branchMap = new HashMap<>();
+    private SnapshotDependencies getSnapshotDependencies(String buildId) {
+        String path = "app/rest/buildTypes/id:" + buildId + "/snapshot-dependencies";
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: this is even duplicated, so please move it to separate method

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 25]
```diff
@@ -101,29 +140,81 @@ public void triggerBuild(JsonObject prPayload) {
...
 
-        for (String mapping : mappings.split(",")) {
-            String[] parts = mapping.split("=>");
-            branchMap.put(parts[0].trim(), parts[1].trim());
+        try {
+            response = getRequestBuilder(target).get();
+            log.debug("Getting JSON with snapshot dependencies from Teamcity. Status: " + response.getStatus());
+
+            String json = response.readEntity(String.class);
+            snapshotDependencies = JsonbBuilder.create().fromJson(json, SnapshotDependencies.class);
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]: and new instance of Jsonb again

**Comment by khermano** [comment]
> @khermano looks good. Two additional points to what I've added inline:
> 
> 1/ Each file must have a license header. Check other existing classes.
> 2/ Does this require some special setup of TeamCity? If yes, do we have it documented?

About point number 2: https://github.com/khermano/tyr/blob/master/tyr-runner/src/test/running_teamcity_test_instance.adoc ...this is all we have for now

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 26]
```diff
@@ -101,29 +146,80 @@ public void triggerBuild(JsonObject prPayload) {
...
+            }
+        }
+
+        String snapshotDependenciesString = jsonb.toJson(snapshotDependencies);
+        Entity<String> json = Entity.json(snapshotDependenciesString);
+        Response response = null;
+
+        try {
+            response = getSnapshotDependenciesRequestBuilder(buildId).put(json);
+            log.debug("Sending JSON with snapshot dependencies in Teamcity. Status: " + response.getStatus());
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]:
```suggestion
            log.debug("Sent JSON with snapshot dependencies to Teamcity. Status: " + response.getStatus());
```

File: tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java [hunk 27]
```diff
@@ -101,29 +146,80 @@ public void triggerBuild(JsonObject prPayload) {
...
+    private SnapshotDependencies getSnapshotDependencies(String buildId) {
+        Response response = null;
+        SnapshotDependencies snapshotDependencies;
 
-        for (String mapping : mappings.split(",")) {
-            String[] parts = mapping.split("=>");
-            branchMap.put(parts[0].trim(), parts[1].trim());
+        try {
+            response = getSnapshotDependenciesRequestBuilder(buildId).get();
+            log.debug("Getting JSON with snapshot dependencies from Teamcity. Status: " + response.getStatus());
```
Comment (xstefank) [inline tyr-runner/src/main/java/org/jboss/tyr/ci/TeamCityCI.java]:
```suggestion
            log.debug("Got JSON with snapshot dependencies from Teamcity. Status: " + response.getStatus());
```

**Comment by xstefank** [comment]
@khermano thanks!

**Comment by khermano** [comment]
np :) thank you
